hoomd/md/IntegratorTwoStep.cc

```diff
@@ -264,29 +268,44 @@ void IntegratorTwoStep::initializeIntegrationMethods()
     }
 
 /*! \param group Group over which to count degrees of freedom.
-    IntegratorTwoStep totals up the degrees of freedom that each integration method provide to the group.
-    Three degrees of freedom are subtracted from the total to account for the constrained position of the system center of
-    mass.
+
+    IntegratorTwoStep totals up the degrees of freedom that each integration method provide to the
+    group.
+
+    When the user has only one momentum conserving integration method applied to the all group,
+    getNDOF subtracts n_dimensions degrees of freedom from the system to account for the pinned
+    center of mass. When the query group is not the group of all particles, spread these these
+    removed DOF proportionately so that the results given by one ComputeThermo on the all group are
+    consitent with the average of many ComputeThermo's on disjoint subset groups.
 */
-unsigned int IntegratorTwoStep::getNDOF(std::shared_ptr<ParticleGroup> group)
+Scalar IntegratorTwoStep::getTranslationalDOF(std::shared_ptr<ParticleGroup> group)
     {
-    int res = 0;
+    // proportionately remove n_dimensions DOF when there is only one momentum conserving
+    // integration method
+    Scalar periodic_dof_removed = 0;
+    if (group->getNumMembersGlobal() == m_pdata->getNGlobal() &&
+        m_methods.size() == 1 &&
+        m_methods[0]->isMomentumConserving())
+        {
+        periodic_dof_removed = Scalar(m_sysdef->getNDimensions()) *
+                               (Scalar(group->getNumMembersGlobal())
+                               / Scalar(m_pdata->getNGlobal()));
```

Comment:
Isn't this by definition 1?

Reply:
Only when the group consists of all particles. HOOMD v2 incorrectly removes N_dimensions DOF from every group.

When the group is only a subset of all particles, this spreads the N_dimensions fixed degrees of freedom out proportionately. With this change if you compute the kinetic temperature of two halves of the system separately and then average them together you will get the same result as if you had computed the kinetic temperature on the all particles. There might be a better way to do this, but I can't think of anything else that would be self-consistent with a simulation of otherwise identical particles.

I also added a check to only remove these DOF when using a momentum conserving integrator. Now the kinetic temperature will be computed correctly when using Langevin where the center of mass of the system moves.